### PR TITLE
[Python] append column with a in-memory Pyarrow Table

### LINKED ISSUE #320: Append column takes a already-computed array.
Allow users to provide an array for `Dataset.append_columns(arr)`.

### PR DESCRIPTION
**Pull request by eddyxu**
Closes #320

### CODE REVIEW

File: python/lance/_lib.pyx [hunk 1]
```diff
@@ -284,23 +290,33 @@ cdef class FileSystemDataset(Dataset):
...
+    def _append_column_with_table(self, table: pyarrow.Table, on: str) -> FileSystemDataset:
+        cdef shared_ptr[CTable] c_table = pyarrow_unwrap_table(table)
+        cdef shared_ptr[CLanceDataset] dataset = GetResultValue(self.lance_dataset.AddColumns(c_table, tobytes(on)))
+        return FileSystemDataset.wrap(static_pointer_cast[CDataset, CLanceDataset](dataset))
+
     def append_column(
         self,
-        field: Field,
-        value: Union[Callable[[pyarrow.Table], pyarrow.Array], Expression],
+        value: Union[Callable[[pyarrow.Table], pyarrow.Array], Expression, pyarrow.Table],
```
[the remark this answers is not in the document]
Comment (changhiskhan) [reply python/lance/_lib.pyx:300]: Should we create a `dataset.merge` method? For now, limit it to horizontal concat. But later on we can allow for appending rows and columns in one shot? Pandas has this functionality, though the pandas merge API is way too complicated. wdyt?

**Review by changhiskhan — APPROVED** [review]
pending #337

File: python/lance/_lib.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/lance/_lib.pyx:300]: ok, i will make this a separate Merge function.

File: python/lance/tests/test_schema_evolution.py [hunk 2]
```diff
@@ -46,7 +49,9 @@ def value_func(x: pa.Table):
         assert x.column_names == ["a"]
         return pa.array([str(i) for i in x.column("a")])
 
-    new_dataset = dataset.append_column(pa.field("c", pa.utf8()), value_func, columns=["a"])
+    new_dataset = dataset.append_column(
+        value_func, field=pa.field("c", pa.utf8()), columns=["a"]
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/lance/tests/test_schema_evolution.py:53]: How about change `columns` to `select` ?

**Pull request merged**